File: examples/proxy.rs
//! Example: simple L4 proxy

use std::{env, io, process};

use tokio::net::{TcpListener, TcpStream};

#[tokio::main(flavor = "current_thread")]
// #[tokio::main]
async fn main() -> io::Result<()> {
    println!("PID is {}", std::process::id());

    let (listen_addr, upstream_addr) = parse_args();

    use tracing::level_filters::LevelFilter;
    use tracing_subscriber::layer::SubscriberExt;
    use tracing_subscriber::util::SubscriberInitExt;
    use tracing_subscriber::{EnvFilter, Layer};

    let (w, _g) = tracing_appender::non_blocking(io::stdout());
    let fmt_layer = tracing_subscriber::fmt::layer().with_writer(w).with_filter(
        EnvFilter::builder()
            .with_default_directive(LevelFilter::DEBUG.into())
            .from_env_lossy()
            .add_directive("otel::tracing=trace".parse().unwrap())
            .add_directive("h2=error".parse().unwrap())
            .add_directive("tower=error".parse().unwrap())
            .add_directive("hyper=error".parse().unwrap()),
    );

    tracing_subscriber::registry()
        .with(fmt_layer)
        // .with(console_subscriber::spawn())
        .init();

    let worker_threads = tokio::runtime::Handle::current().metrics().num_workers();
    let pipe_size = splicer::pipe::Pipe::new()?.size();
    println!(
        "tokio worker threads: {worker_threads}, splice pipe size: {pipe_size} bytes (per direction, 2 per connection)"
    );

    tokio::select! {
        res = serve(listen_addr, upstream_addr) => {
            if let Err(err) = res {
                eprintln!("Serve failed {err}");
            }
        }
        _ = tokio::signal::ctrl_c() => {
            println!("Received Ctrl + C, shutting down");
        }
    }

    Ok(())
}

fn parse_args() -> (String, String) {
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("usage: {} <listen_port> <upstream_port>", args[0]);
        process::exit(2);
    }
    let listen_port: u16 = args[1].parse().expect("invalid listen port");
    let upstream_port: u16 = args[2].parse().expect("invalid upstream port");
    (
        format!("0.0.0.0:{listen_port}"),
        format!("127.0.0.1:{upstream_port}"),
    )
}

async fn serve(listen_addr: String, upstream_addr: String) -> io::Result<()> {
    let listener = TcpListener::bind(&listen_addr).await?;

    loop {
        let (incoming, remote_addr) = match listener.accept().await {
            Ok(accepted) => accepted,
            Err(e) if e.kind() == io::ErrorKind::ConnectionAborted => {
                println!("Connection aborted.");
                continue;
            }
            Err(e) => {
                eprintln!("Failed to accept: {e:#?}");
                break Err(e);
            }
        };

        println!("Process incoming connection from {remote_addr}");

        tokio::spawn(forwarding(incoming, upstream_addr.clone()));
    }
}

async fn forwarding(mut stream1: TcpStream, upstream_addr: String) -> io::Result<()> {
    let stream2 = TcpStream::connect(&upstream_addr).await;

    let mut stream2 = match stream2 {
        Ok(s) => s,
        Err(e) => {
            eprintln!("Failed to connect to remote server: {e}");
            return Err(e);
        }
    };

    let instant = std::time::Instant::now();

    let io_sl2sr = splicer::splice::SpliceCtx::new()?.into();
    let io_sr2sl = splicer::splice::SpliceCtx::new()?.into();

    let traffic = splicer::io::SpliceBidiIo { io_sl2sr, io_sr2sl }
        .execute(&mut stream1, &mut stream2)
        .await;

    let total = traffic.sum();
    let cost = instant.elapsed();
    println!(
        "Forwarded traffic: total: {total} B, time: {:.2} s, avg: {:.4} B/s, error: {:?}",
        cost.as_secs_f64(),
        total as f64 / cost.as_secs_f64(),
        traffic.error
    );

    Ok(())
}

File: examples/proxy.go
package main

import (
	"context"
	"fmt"
	"io"
	"log"
	"net"
	"os"
	"os/signal"
	"runtime"
	"strconv"
	"sync"
	"syscall"
	"time"
)

func main() {
	fmt.Printf("PID is %d\n", os.Getpid())

	// Go creates splice pipes internally and doesn't expose them; it requests
	// 1 MiB (internal/poll.maxSpliceSize), but the kernel caps that at
	// fs.pipe-max-size (and stays at the 64 KiB default if the request fails).
	var pipeMax int
	if b, err := os.ReadFile("/proc/sys/fs/pipe-max-size"); err == nil {
		fmt.Sscan(string(b), &pipeMax)
	}
	fmt.Printf("GOMAXPROCS: %d, splice pipe: requests %d bytes, fs.pipe-max-size = %d bytes (per direction, 2 per connection)\n",
		runtime.GOMAXPROCS(0), 1<<20, pipeMax)

	listenAddr, upstreamAddr := parseArgs()

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		if err := serve(ctx, listenAddr, upstreamAddr); err != nil {
			log.Printf("Serve failed: %v", err)
		}
	}()

	<-sigChan
	fmt.Println("Received Ctrl + C, shutting down")
	cancel()

	time.Sleep(100 * time.Millisecond)
}

func parseArgs() (string, string) {
	if len(os.Args) != 3 {
		fmt.Fprintf(os.Stderr, "usage: %s <listen_port> <upstream_port>\n", os.Args[0])
		os.Exit(2)
	}
	listenPort, err := strconv.ParseUint(os.Args[1], 10, 16)
	if err != nil {
		fmt.Fprintf(os.Stderr, "invalid listen port: %v\n", err)
		os.Exit(2)
	}
	upstreamPort, err := strconv.ParseUint(os.Args[2], 10, 16)
	if err != nil {
		fmt.Fprintf(os.Stderr, "invalid upstream port: %v\n", err)
		os.Exit(2)
	}
	return fmt.Sprintf("0.0.0.0:%d", listenPort), fmt.Sprintf("127.0.0.1:%d", upstreamPort)
}

func serve(ctx context.Context, listenAddr, upstreamAddr string) error {
	listener, err := net.Listen("tcp", listenAddr)
	if err != nil {
		return fmt.Errorf("failed to listen on %s: %w", listenAddr, err)
	}
	defer listener.Close()

	fmt.Printf("Listening on %s\n", listenAddr)

	for {
		select {
		case <-ctx.Done():
			return nil
		default:
		}

		if tcpListener, ok := listener.(*net.TCPListener); ok {
			tcpListener.SetDeadline(time.Now().Add(100 * time.Millisecond))
		}

		conn, err := listener.Accept()
		if err != nil {
			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
				continue
			}
			if netErr, ok := err.(net.Error); ok && netErr.Temporary() {
				log.Printf("Temporary accept error: %v", err)
				continue
			}
			return fmt.Errorf("failed to accept: %w", err)
		}

		remoteAddr := conn.RemoteAddr()
		fmt.Printf("Process incoming connection from %s\n", remoteAddr)

		go forwarding(conn, upstreamAddr)
	}
}

func forwarding(stream1 net.Conn, upstreamAddr string) error {
	defer stream1.Close()

	stream2, err := net.Dial("tcp", upstreamAddr)
	if err != nil {
		log.Printf("Failed to connect to remote server: %v", err)
		return err
	}
	defer stream2.Close()

	startTime := time.Now()

	result, err := copyBidirectional(stream1, stream2)

	elapsed := time.Since(startTime)

	fmt.Printf("Forwarded traffic: %+v, avg: %.4f B/s\n", result, float64((result.BytesForward+result.BytesReverse))/elapsed.Seconds())

	if err != nil {
		log.Printf("Failed to copy data: %v", err)
		return err
	}

	return nil
}

type TrafficStats struct {
	BytesForward uint64
	BytesReverse uint64
}

func (t TrafficStats) String() string {
	return fmt.Sprintf("TrafficStats { bytes_forward: %d, bytes_reverse: %d }",
		t.BytesForward, t.BytesReverse)
}

func copyBidirectional(conn1, conn2 net.Conn) (*TrafficStats, error) {
	var stats TrafficStats
	var wg sync.WaitGroup
	var err1, err2 error

	wg.Add(2)

	go func() {
		defer wg.Done()
		defer conn2.Close()
		n, err := io.Copy(conn2, conn1)
		stats.BytesForward = uint64(n)
		err1 = err
	}()

	go func() {
		defer wg.Done()
		defer conn1.Close()
		n, err := io.Copy(conn1, conn2)
		stats.BytesReverse = uint64(n)
		err2 = err
	}()

	wg.Wait()

	if err1 != nil && err1 != io.EOF {
		return &stats, err1
	}
	if err2 != nil && err2 != io.EOF {
		return &stats, err2
	}

	return &stats, nil
}
